### ISSUE TITLE
Dictionary unpacking interferes with lambda expression type inference

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1724]
**Describe the bug**
When the return value of a lambda expression is a dictionary into which known-type dictionaries are unpacked, Pyright is not able to infer the type of the dictionary values.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That `bar` would be `dict[str, str]`.

**Screenshots or Code**

```python
# pyright: strict

from functools import reduce

dicts = [{'a': 'b'}, {'c': 'd'}]

foo = reduce(lambda x, y: x | y, dicts)
reveal_type(foo)  # Type of "foo" is "dict[str, str]"

# Argument type is partially unknown
#   Argument corresponds to parameter "function" in function "reduce"
#   Argument type is "(x: dict[Unknown, Unknown], y: dict[Unknown, Unknown]) -> dict[Unknown, Unknown]"
# [...]
bar = reduce(lambda x, y: {**x, **y}, dicts)
reveal_type(bar)  # Type of "bar" is "dict[Unknown, Unknown]"

```

**VS Code extension or command-line**
Either, 1.1.128

**Comment by erictraut** [comment]
The problem in this case was that the initial solution for `_T` in the `reduce` method was `dict[Unknown, Unknown]`, but this should have later been refined to `dict[str, str]`. The constraint solver stuck with the original solution rather than replacing it with the preferable one. I've tweaked a heuristic in the constraint solver to prefer types that contain no "Unknown" element. 

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.129, which I just published. It will also be included in the next release of pylance.
